src/kernl/optimizer/cuda_graph.py

```diff
@@ -12,48 +12,74 @@
 #  See the License for the specific language governing permissions and
 #  limitations under the License.
 #
-from collections import defaultdict, deque
+import math
 from typing import Callable, Union
 
 import torch
-import triton
 from torch._dynamo import utils as dynamo_utils
 from torch._inductor.compile_fx import cudagraphify_impl
 from torch._subclasses import FakeTensor
 
+from kernl.optimizer.pool_cuda_graphs import CudaGraphPool, get_aligned_size
 
-static_inputs_pool: dict[(int, torch.dtype), deque[torch.Tensor]] = defaultdict(deque)
 
+static_inputs_pool = []
 
-def get_static_inputs(model_inputs: list[torch.Tensor]) -> list[torch.Tensor]:
-    """
-    Copy input tensors to a managed pool of tensors (to limit memory footprint).
-    All tensors need to be declared as static inputs to CUDA graph.
 
-    @param model_inputs: list of inputs
-    @return: list of inputs to be used in CUDA graphs
+def prepare_inputs(inputs: list[torch.Tensor], pools: list[CudaGraphPool]) -> list[torch.Tensor]:
+    """
+    Copy the inputs in the CUDA graphs memory pool and return tensor copies.
+    Follows a greedy bin packing algorithm (first-fit decreasing) to minimize the number of pools:
+    - sort the items in decreasing order of size ;
+    - insert them one by one into the first bin that has room for it.
+
+    :param inputs: list of tensors to copy in the pool
+    :param pools: list of available pools
+    :return: list of tensors that are copies of the inputs in the pool
```

Comment:
"copy of input tensors with the underlying storage in the memory pool" or something like this maybe, not sure. I think the comment is ambiguous

Reply:
updated